### PR TITLE
Avro Event Format for CloudEvents

### LINKED ISSUE #426: Avro format enconding
I would like to propose an official Avro encoding schema to be supported by CloudEvents.

I've seen a lot of uses cases, in special here in my current project, with Apache Kafka and the many of them were using Avro as the format. Inspired by [this](https://gist.github.com/rbramley/c00bbd95f3f6ecc0cc1f90e5d6d6c90c) Gist (thanks @rbramley), the following schema could be the first draft:

```json
{
  "namespace":"io.cloudevents",
  "type":"record",
  "name":"CloudEvent",
  "version": "1",
  "doc": "CloudEvents Spec for Avro format encoding (the version is just for the schema itself)",
  "fields":[
    {
      "name":"specversion",
      "type":"string"
    },
    {
      "name":"type",
      "type":"string"
    },
    {
      "name":"source",
      "type":"string"
    },
    {
      "name":"id",
      "type":"string",
    },
    {
      "name":"time",
      "type":[
        "null",
        "int"
      ],
      "logicalType": "date"
    },
    {
      "name":"schemaurl",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"contenttype",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"data",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"extensions",
      "type":[
        "null",
        {
          "type":"map",
          "values":[
            "string",
            "int",
            "long",
            "float",
            "double",
            "boolean",
            "bytes"
          ]
        }
      ]
    },
  ]
}
```

- Official Avro spec available [here](https://avro.apache.org/docs/1.8.2/spec.html)

__Extensions__

To define a top-level extension (like below) is not allowed by Avro spec for `avsc`, because we do not have the feature for dynamic fields like JSON. Because of this, you will see the `extension` attribute to the example above.

```json
...
    "comexampleextension1" : "value",
    "comexampleextension2" : {
        "othervalue": 5
    }
...
```

For that, we may employ Avro spec for `ìdl` that is a higher-level language for authoring Avro schemas.

- Official Avro IDL docs available [here](https://avro.apache.org/docs/1.8.2/idl.html)

__Special Types__

What I mean with _special types_, logical types in JSON Schema, are: `uri-reference`. We do not have this type in Avro.

For that, again, we may employ Avro IDL.

### PR DESCRIPTION
**Pull request by fabiojose**
This PR closes #426.

This is the first version, please comment and improve!

### CODE REVIEW

File: avro-format.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# Avro Event Format for CloudEvents
+
+## Abstract
+
+The Avro Format for CloudEvents defines how events attributes are expressed in the [Avro 1.9.0 Specification][avro-spec]
```
Comment (gunnarmorling) [inline avro-format.md]:
```suggestion
The Avro Format for CloudEvents defines how events attributes are expressed in the [Avro 1.9.0 Specification][avro-spec].
```

File: avro-format.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are mapped to the Avro
+type-system. This specification explicitly map each attribute.
```
Comment (gunnarmorling) [inline avro-format.md]:
```suggestion
type-system. This specification explicitly maps each attribute.
```

File: avro-format.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+| Any           | See [2.2](#22-mapping-any-typed-attributes)                            |
+| URI-reference | [string][avro-primitives] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [string][avro-primitives] following [RFC 3339][rfc3339] (ISO 8601)     |
+
+Extension specifications MAY define diverging mapping rules for the values of
+attributes they define.
+
+### 2.2 Mapping Any-typed Attributes
+
+`Any`-typed CloudEvents values can either hold a `String`, or a `Binary` value,
```
Comment (gunnarmorling) [inline avro-format.md]: It could be all the other types.

File: avro-format.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+or a `Map`. Avro type system SHOULD NOT satisfy the map-with-maps in a recursive
```
Comment (gunnarmorling) [inline avro-format.md]: What's the motivation for discouraging maps of maps?

File: avro-format.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+Users of Avro MUST use a message whose binary encoding is identical
+to the one described by the [CloudEvent Avro Schema](./spec.avsc):
+
+```json
+{
+  "namespace":"io.cloudevents",
+  "type":"record",
+  "name":"CloudEvent",
+  "version":"0.2",
+  "doc":"Avro Event Format fo CloudEvents",
```
Comment (duglin) [inline avro-format.md]: s/fo/of/

**Comment by duglin** [comment]
@fabiojose can you sign your commits?

File: avro-format.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
+# Avro Event Format for CloudEvents
+
+## Abstract
+
+The Avro Format for CloudEvents defines how events attributes are expressed in the [Avro 1.9.0 Specification][avro-spec].
```
Comment (duglin) [inline avro-format.md]: please wrap things at 80 columns

File: avro-format.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+The CloudEvents type system MUST be mapped to Avro types as follows.
+
+| CloudEvents   | Avro                                                                   |
+| ------------- | ---------------------------------------------------------------------- |
+| Integer       | [int][avro-primitives]                                                 |
+| String        | [string][avro-primitives]                                              |
+| Binary        | [bytes][avro-primitives]                                               |
+| Map           | [map][avro-primitives]                                                 |
+| Any           | See [2.2](#22-mapping-any-typed-attributes)                            |
```
Comment (duglin) [inline avro-format.md]: I like that the order is the same as what's in the spec... but ANY has been moved to the end now  :-)

File: avro-format.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+`Any`-typed CloudEvents values can either hold a `String`, or a `Binary` value,
+or a `Map`. Avro type system SHOULD NOT satisfy the map-with-maps in a recursive
+way, i.e., every `map` declaration REQUIRE the `values` with primitives types
+or a declared type.
+
+### 2.3 Optional Attributes
+
+CloudEvents Spec defines optional attributes. The Avro format defines that
+every optional field MUST use the `null` type and the actual type through
+the [union][avro-unions].
```
Comment (duglin) [inline avro-format.md]: Does this mean that optional attributes will always appear, just with a "null" value? Is it a "MUST appear" or "MAY appear"? What happens if the send doesn't include it - are they violating something?

File: avro-format.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+Users of Avro MUST use a message whose binary encoding is identical
+to the one described by the [CloudEvent Avro Schema](./spec.avsc):
+
+```json
+{
+  "namespace":"io.cloudevents",
+  "type":"record",
+  "name":"CloudEvent",
+  "version":"0.2",
```
Comment (duglin) [inline avro-format.md]: 0.3

**Comment by duglin** [comment]
@fabiojose can you address the comments before Tuesday EOD?

File: avro-format.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
+# Avro Event Format for CloudEvents
```
Comment (duglin) [inline avro-format.md]: For consistency, please add `  - Version 0.4-wip` to the tile.

**Comment by NickLarsenNZ** [comment]
@fabiojose, one of the checks failed (commit signoffs). See details: https://github.com/cloudevents/spec/pull/438/checks?check_run_id=145188904

**Comment by duglin** [comment]
@fabiojose you around?

File: avro-format.md (hunk 4 above)
Comment (fabiojose) [reply avro-format.md]: @gunnarmorling It's possible, but the map inside of another map needs the type declaration. It is different from JSON maps.

Let's see an example:

```json
{
  "name":"metadata",
  "type":{
    "type":"map",
    "values":[
      "null",
      "int",
      "float",
      "string",
      "boolean",
      "long",
      {
        "type":"map",
        "values":[
          "null",
          "int",
          "float",
          "string",
          "boolean",
          "long"
        ]
      }
    ]
  }
}
```
If I want another `map` I need to declare. There is no way to declare a `Any`-typed map.

File: avro-format.md (hunk 8 above)
Comment (fabiojose) [reply avro-format.md]: @duglin it is "MAY appear". Using the `null` means that an optional attribute appears if it has value or no appear if there is no value at all.

**Pull request closed without merging**